index.src.html

```diff
@@ -819,93 +819,94 @@ <h3 id=interface-reporting-observer>Interface {{ReportingObserver}}</h3>
-     the <a>registered reporting observer list</a> of its
-     associated <a>environment settings object</a>.
+  The <dfn method for=ReportingObserver><code>observe()</code></dfn>
+  method, when invoked, must run these steps:
 
-  4. If the <a>context object's</a> {{futureOnly}} <a>option</a> is true, stop.
+  1. Let |settings| be the <a>environment settings object</a> associated with
+     the <a>context object</a>.
 
-  5. For each <a>report</a> |report| in the <a>report buffer</a> associated with
-     |settings|, execute [[#add-report]] with |report| and the <a>context object</a>.
+  2. Append the <a>context object</a> to the <a>registered reporting observer
+     list</a> of its associated <a>environment settings object</a>.
 
-  6. <a>Queue a task</a> to [[#invoke-observers]] with a list containing the <a>context object</a>
+  3. If the <a>context object</a>'s {{futureOnly}} <a>option</a> is true, return.
+
+  4. Set <a>context object</a>'s {{futureOnly}} <a>option</a> to true.
+
+  5. For each |report| in the <a>report buffer</a> associated with |settings|,
+     execute [[#add-report]] with |report| and the <a>context object</a>.
 
   The <dfn method for=ReportingObserver><code>disconnect()</code></dfn> method,
   when invoked, must run these steps:
 
-  1. If the <a>context object</a> is not <a>registered</a>, stop.
+  1. If the <a>context object</a> is not <a>registered</a>, return.
 
   2. Remove the <a>context object</a> from the <a>registered reporting observer
      list</a> of its associated <a>environment settings object</a>.
 
   <h3 id="add-report" algorithm>
-    Add |report| to reporting |observer|
+    Add |report| to |observer|
   </h3>
 
-  This algorithm adds |report| as a {{Report}} object to |observer|'s <a>report
-  list</a>, so long as |report|'s <a>type</a> is observable by |observer|.
+  Given a <a>report</a> |report| and a {{ReportingObserver}} |observer|, this
+  algorithm adds |report| to |observer|'s <a>report list</a>, so long as
+  |report|'s <a>type</a> is observable by |observer|.
 
   1. If |observer|'s <a>options</a> has a {{ReportingObserverOptions/types}}
-     member which does not contain |report|'s [=report/type=], stop.
+     member which does not contain |report|'s [=report/type=], return.
 
   2. Create a new {{Report}} |r| with {{Report/type}} initialized to |report|'s
      [=report/type=], {{Report/url}} initialized to |report|'s [=report/url=],
      and {{Report/body}} initialized to |report|'s [=report/body=].
 
   Issue: how to polymorphically initialize body?
 
-  3. Add |r| to the end of |observer|'s <a>report list</a>.
-      
+  3. Append |r| to |observer|'s <a>report list</a>.
+
+  4. If the size of |observer|'s <a>report list</a> is 1, <a>Queue a task</a> to
+     [[#invoke-observers]] with a copy of the <a>registered reporting observer
+     list</a> of the <a>environment settings object</a> associated with
+     |observer|.
+
   <h3 id="notify-observers" algorithm>
```

Comment:
This algorithm feels like it's in the wrong spot... right now 5.1 calls 5.2, which calls 5.4. Then 4.1 calls this one (5.3) which also calls 5.2. I'm not sure where it should go, but maybe switching with `#add-report` would make sense? Then the calls would always flow 'down' in the document.

Reply:
That makes sense. I've swapped them.